### PR TITLE
New Rule: validateCommentPosition

### LINKED ISSUE #651: Rule to validate the positioning of comments about a statement
### validateCommentPosition

Type: `'String'`

Values:

* 'above'
* 'beside'

### Example

`validateCommentPosition: 'above'`

Valid:

```js
// This is a valid comment
1 + 1;
```

Invalid:
```js
1 + 1; // An invalid comment
```

`validateCommentPosition: 'beside'`

Valid:

```js
1 + 1; // A valid comment
```

Invalid:
```js
// This is an invalid comment
1 + 1;
```

If you'd like to support both styles, then don't supply a value to this rule.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/4696455-rule-to-validate-the-positioning-of-comments-about-a-statement?utm_campaign=plugin&utm_content=tracker%2F281640&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F281640&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by bjdixon**
Fixes #651

### CODE REVIEW

**Comment by markelog** [comment]
How about `/**/`?

**Comment by bjdixon** [comment]
I guess I assumed this was for line comments. I'll add block comments if that's desired. Let me know.

Also, I had a question. For the 'beside' style this is fine:
``` 
// jscs: disable 
1 + 1;
``` 
but should enable trigger an error or pass?
```
// jscs: enable
1 + 1;
```

**Comment by markelog** [comment]
> I guess I assumed this was for line comments.

Yeah, i was asking if you ignore them or not

> but should 'enable' trigger an error or pass?

i'd say just ignore them. I guess we would need to expand on other constructions as well? See discussion in https://github.com/jscs-dev/node-jscs/issues/1685

**Comment by bjdixon** [comment]
I've added some tests showing that block comments are ignored. I've also added an array to ultils.js to filter out inline configurations and an allExcept parameter to this rule.

**Comment by markelog** [comment]
LGTM

File: lib/rules/validate-comment-position.js [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+ * Comments that start with the following keywords are also ignored:
+ * `eslint`, `jshint`, `jslint`, `istanbul`, `global`, `exported`, `jscs`, `falls through`
+ * eg. // jshint strict: true
+ *
+ * Type: `Object`
+ *
+ * Value:
+ *
+ * - `Object`:
+ *    - `mode`: `above` or `beside`
```
Comment (mrjoelkemp) [inline lib/rules/validate-comment-position.js]: Why not just call it `position`?

File: lib/rules/validate-comment-position.js [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+    getOptionName: function() {
+        return 'validateCommentPosition';
+    },
+
+    check: function(file, errors) {
+
+        function isExcepted(comment) {
+            var trimmedComment = comment.trim();
+            for (var i = 0; i < exceptions.length; i++) {
```
Comment (mrjoelkemp) [inline lib/rules/validate-comment-position.js]: I was confused as to where `exceptions` was defined. Seems like that var declaration `var exceptions = this._exceptions;` could be moved up above this helper.

Comment (mrjoelkemp) [inline lib/rules/validate-comment-position.js]: Also consider just using `return exceptions.some(...`

File: lib/rules/validate-comment-position.js [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    },
+
+    getOptionName: function() {
+        return 'validateCommentPosition';
+    },
+
+    check: function(file, errors) {
+
+        function isExcepted(comment) {
+            var trimmedComment = comment.trim();
```
Comment (mrjoelkemp) [inline lib/rules/validate-comment-position.js]: Why do you need to trim?

File: test/specs/rules/validate-comment-position.js [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+            }, assert.AssertionError);
+        });
+    });
+
+    describe('mode value "above"', function() {
+        beforeEach(function() {
+            checker.configure({ validateCommentPosition: { mode: 'above' }});
+        });
+
+        it('should report unexpected position for 1 + 1; // invalid comment', function() {
```
Comment (mrjoelkemp) [inline test/specs/rules/validate-comment-position.js]: Non-blocker: we typically don't put code in the descriptions. A general description of the code would suffice. Like `it('should report on a comment beside a statement', ...`

**Comment by mrjoelkemp** [comment]
How about a test for:

```
/* my sweet comment */
/* my super sweet comment */
```

or any combination of stacked comments

**Comment by mrjoelkemp** [comment]
Nice work here @bjdixon!

**Comment by bjdixon** [comment]
@mrjoelkemp 

position is a much better description than mode. I've made that change throughout.

I've manually hoisted the variable declarations in the check function. Hopefully that's more readable now.

I used trim and didn't use ```exceptions.some(...)``` because I wanted to skip comments that begin with the keywords (ignoring whitespace). ```trimmedComment.indexOf(exceptions[i]) >= 0``` was a bug that I've corrected to ```trimmedComment.indexOf(exceptions[i]) === 0```. All the eslint, jshint etc pragmas begin with the keywords listed. I think that skipping comments if they just contain keywords may be too permissive. eg. ```// I really like global warming``` isn't a valid pragma but ```// global var1, var2 ``` is. Let me know if you disagree with this approach.

I've made changes to the test descriptions to (hopefully) be in keeping with the style you suggested. I also added a test for stacked comments.

**Comment by mrjoelkemp** [comment]
Woot! Looks great @bjdixon. Thanks for the effort.

@markelog mind giving this another glance?

**Comment by zxqfox** [comment]
Looks good. Is this the first rule without boolean value? ;-)

**Comment by alawatthe** [comment]
I had a look at this pr while looking at https://github.com/jscs-dev/node-jscs/issues/1685. One thing I noted was, that you are checking, if the comment starts with one of the ```inlineConfigurationKeywords``` meaning, that a comment like

```js
// globalize this function
```
will be recognized as pragma (since "global" is one of the ```inlineConfigurationKeywords```), although it is not a pragma.

I would suggest something like

```js
 if (trimmedComment.split(/( |:|-)/)[0] === exceptions[i]) {
```

Yes, I know, this does not work with "falls through", but for this we could introduce a special case.

**Comment by markelog** [comment]
Sounds like we should move pragma checker to `utils` not only the list alone

**Comment by bjdixon** [comment]
@markelog Yeah, I think having the pragma checker function in utils will remove some duplication.

@alawatthe good point on words that begin comments that could be made up of keywords + other characters. I implemented using a small concatenation instead of the regex + extra special case check as it's faster and should be easier to maintain. If the consensus is to use regex though I'm happy to do that.

**Comment by markelog** [comment]
Okay, added one more test so coveralls wouldn't swear

**Pull request closed without merging**